board: arm64: Add suport for ROC-RK3568-PC

cocoeoli opened the pull request:
This is support for AArch64 development board.
The board uses 4-core Cortex-A55, which are based on the ARMv8.2 architecture.
In addition,we support smp support and
it can use 4-cores to run basic samples.
Associated PR: #63343

carlocaione commented on soc/arm64/rockchip/rk3568/soc.c:
```diff
@@ -0,0 +1,18 @@
+#ifdef CONFIG_SOC_RK3568
+
+#ifndef CONFIG_PM_CPU_OPS_PSCI
+int pm_cpu_on(unsigned long cpuid, uintptr_t entry_point)
```
why is this needed? If `CONFIG_PM_CPU_OPS_PSCI` is not defined, `pm_cpu_on` should return `-ENOTSUP` (if `CONFIG_PM_CPU_OPS` is defined).

cocoeoli replied:
Thanks, it seem not necessary for specific scenarios. I think there may be situations where `CONFIG_PM_CPU_OPS_PSCI` is not defined but `pm_cpu_on` is used.

carlocaione commented on arch/arm64/core/reset.c:
```diff
@@ -125,6 +125,11 @@ void z_arm64_el2_init(void)
 
 	reg = read_hcr_el2();
+	/* when EL2 is enable in current security status:
+	 * Clear TGE bit: All exceptions that would not be routed to EL2;
+	 * Clear IMO bit: Physical IRQ interrupts are not taken to EL2 and EL3.
```
```suggestion
	 * Clear TGE bit: All exceptions that would be routed to EL2 are routed to EL1;
	 * Clear IMO bit: Physical IRQ interrupts are not taken to EL2 and EL3.
```

I'm a bit lost on the IMO bit. According to the documentation: `If the value of HCR.TGE is 0, then virtual SError interrupts are enabled in Non-secure state.`. Where did you get that comment regarding physical IRQs?

cocoeoli replied:
In the Arm® Architecture Reference Manual(page D13-3336), the HCR_EL2 register's IMO bit has below feature: when IMO bit is 1, `Physical IRQ interrupts are taken to EL2, unless they are routed to EL3.`. 
![image](https://github.com/zephyrproject-rtos/zephyr/assets/40254934/4e9e3de1-6aed-4374-b050-d4581a413a79)

What you describe may be the HCR_ELR.AMO bit:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/40254934/4f0cc3d0-85bb-45c4-b6b7-a2f5dc7ff1ef)
So, I clear HCR_EL2.IMO bit.

povergoing replied:
Then do we need also to clear the AMO bit?
Also, it would be better to separate the changes in `reset.c` and `cpu.h` into a new commit.

cocoeoli replied:
Thanks,  the reason why there is no fatal now may be that no SError occurred.  But when zephyr boot using uboot on roc_rk3568_pc, HCR_EL2.AMO bit are set to 1 too.  Clear it may be a good choice. 
On some platforms such as qemu_ cortex_ a53, this operation is not necessary, because the value of HCR_EL2 register is zero. I will add a new commit to annouce it.

carlocaione reviewed (changes requested):
nit: Can you fix spelling and spacing in the first commit title?

`arch: arm64: Re-init HCR_EL2 in z_arm64_el2_init`

cocoeoli commented on arch/arm64/core/reset.c:
```diff
@@ -125,6 +125,13 @@ void z_arm64_el2_init(void)
+	 * Clear IMO bit: Physical IRQ interrupts are not taken to EL2 and EL3.
+	 */
+	reg &= ~(HCR_IMO_BIT |
+		HCR_AMO_BIT | HCR_TGE_BIT);
```
Ok, I have changed the title.

povergoing commented:
Can you please fix the CI failures?

cocoeoli commented:
> Can you please fix the CI failures?
Some failure are caused by building other board, such as `west twister -p circuitdojo_feather_nrf9160_ns -s tests/kernel/common/kernel.common` will cause below failures:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/40254934/9293f18e-ccc9-49d5-b2fe-b664525d9283)
Could this be considered as an issue with the test of `circuitdojo_feather_nrf9160_ns` board?

cocoeoli commented:
> Can you please fix the CI failures?

Yes, I modified the documentation, it seems work.

cocoeoli commented:
> Can you please fix the CI failures?

I am not very familiar with Github CI. Can you tell me why the build process of other board affects this PR's check?
`west build -p -b qemu_xtensa_mmu tests/kernel/fatal/message_capture -T kernel.logging.message_capture`

carlocaione commented:
@cocoeoli try to rebase on the latest `main`

cocoeoli commented:
> @cocoeoli try to rebase on the latest `main`

Thanks, I will have a try.

cocoeoli committed:
arch: arm64: Re-init HCR_EL2 in z_arm64_el2_init

HCR_EL2 is configured to certain value by some
loaders such as Uboot on some arm64 boards(such as roc_rk3568_pc),
When HCR_EL2.TGE, HCR_EL2.AMO and HCR_EL2.IMO bits are
set to 1, some unpredictable behaviors may occur during
zephyr boot. So we clear these bits to avoid it.

Signed-off-by: Charlie Xiong <1981639884@qq.com>

cocoeoli committed:
boards: arm64: provide support for ROC-RK3568-PC

This is support for AArch64 development board.
The board uses 4-core Cortex-A55, which are based on
the ARMv8.2 architecture.
In addition,we support smp support and
it can use 4-cores to run basic samples.

Signed-off-by: Charlie Xiong <1981639884@qq.com>

cocoeoli commented:
@carlocaione @povergoing Please help me review these commits and see if there are any other issues?

povergoing reviewed (approved):
LGTM

Cherish-Gww commented:
@cocoeoli 
Hi, I have run this patch on my RK3566 platform, that cool, next I want try zvm on my platform.
